### PR TITLE
Generate decision letter

### PR DESCRIPTION
**Pull request by gnott**
Fixes https://github.com/elifesciences/elife-bot/issues/967

Will require settings to be merged prior to this PR https://github.com/elifesciences/builder-private/pull/347

More functions are moved into the `letterparser_provider.py` so they can be used by multiple activities. As a result, in `activity/activity_ValidateDecisionLetterInput.py` the multiple steps to convert a `.zip` into decision letter output is split into two parts: zip -> article objects, and articles to XML.

`GenerateDecisionLetterJATS` is the new activity. It will download the `.zip`, generate XML, upload it to a folder in the decision letter output bucket, and store a couple values into a session for use later when this XML is posted to an API endpoint.

Folder name and file name values are formatted from strings in `settings.py`.

The new activity is added to the `IngestDecisionLetter` workflow so we can try it out.

### CODE REVIEW

File: activity/activity_GenerateDecisionLetterJATS.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        }
+
+        # Load the config
+        self.letterparser_config = letterparser_provider.letterparser_config(self.settings)
+
+    def do_activity(self, data=None):
+        """
+        Activity, do the work
+        """
+        if self.logger:
```
Comment (giorgiosironi) [inline activity/activity_GenerateDecisionLetterJATS.py]: there should always be a `self.logger`; in tests, pass in a mock that just ignores all calls

File: activity/activity_ValidateDecisionLetterInput.py [hunk 2]
```diff
@@ -63,25 +62,25 @@ def do_activity(self, data=None):
...
+            letterparser_provider.process_zip(
+                self.input_file,
+                config=self.letterparser_config,
+                temp_dir=self.directories.get("TEMP_DIR"),
+                logger=self.logger))
 
-        # Convert docx to articles
-        self.statuses["build"], self.articles = letterparser_provider.docx_to_articles(
-            docx_file_name, config=self.letterparser_config, logger=self.logger)
+        self.copy_statuses(statuses)
```
Comment (giorgiosironi) [inline activity/activity_ValidateDecisionLetterInput.py]: this is storing data in `self` so `set_` as a verb can make it more explicit?

File: activity/activity_ValidateDecisionLetterInput.py [hunk 3]
```diff
@@ -63,25 +62,25 @@ def do_activity(self, data=None):
...
 
-        # Validate content of articles
-        self.statuses["valid"], error_messages = letterparser_provider.validate_articles(
-            self.articles, logger=self.logger)
+        # part 2 articles to XML
+        self.xml_string, statuses = letterparser_provider.process_articles_to_xml(
```
Comment (giorgiosironi) [inline activity/activity_ValidateDecisionLetterInput.py]: the logic is reasonably isolated from the activity, which helps with reuse and testability

File: activity/activity_GenerateDecisionLetterJATS.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        self.articles, asset_file_names, statuses, error_messages = (
+            letterparser_provider.process_zip(
+                self.input_file,
+                config=self.letterparser_config,
+                temp_dir=self.directories.get("TEMP_DIR"),
+                logger=self.logger))
+        # part 2 articles to XML
+        self.xml_string, statuses = letterparser_provider.process_articles_to_xml(
+            self.articles,
+            self.directories.get("TEMP_DIR"),
```
Comment (giorgiosironi) [inline activity/activity_GenerateDecisionLetterJATS.py]: Just to check: is this just using the temporary directory to write new files into it, or relying on anything being present there? The latter would be a smell

File: activity/activity_GenerateDecisionLetterJATS.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            self.logger,
+            pretty=True,
+            indent="    ")
```
Comment (giorgiosironi) [inline activity/activity_GenerateDecisionLetterJATS.py]: don't know what this `indent` parameter does but seems weird compared to highly structured input like the article :thinking:

File: activity/activity_GenerateDecisionLetterJATS.py (hunk 4 above)
Comment (gnott) [reply activity/activity_GenerateDecisionLetterJATS.py]: This directory name is passed along to the function in the `letterparser` library, where it is the location to unzip files from a `.zip` file. Outside of that it isn't used much.

File: activity/activity_GenerateDecisionLetterJATS.py (hunk 5 above)
Comment (gnott) [reply activity/activity_GenerateDecisionLetterJATS.py]: The `indent` value is passed to `minidom` when generating the XML output: https://docs.python.org/3/library/xml.dom.minidom.html?highlight=minidom#xml.dom.minidom.Node.toprettyxml  if `pretty` is `True`.  

This happens in `letterparser.generate.output_xml()` (https://github.com/elifesciences/decision-letter-parser/blob/develop/letterparser/generate.py#L232)

File: activity/activity_GenerateDecisionLetterJATS.py (hunk 1 above)
Comment (gnott) [reply activity/activity_GenerateDecisionLetterJATS.py]: This is a holdover from old lettuce tests I think when the logger was not always present. I can remove this 👍

File: activity/activity_ValidateDecisionLetterInput.py (hunk 2 above)
Comment (gnott) [reply activity/activity_ValidateDecisionLetterInput.py]: Sure, it sounds like a good naming option.

Comment (gnott) [reply activity/activity_ValidateDecisionLetterInput.py]: This small function is present in two activities - a tiny bit of duplicated code as a result, but I'm not sure how useful it is for reuse, and if so, where to put it.

Comment (giorgiosironi) [reply activity/activity_ValidateDecisionLetterInput.py]: Very small amount of code so wouldn't worry about reuse

**Pull request merged**